build-leaf/mint-leaf.yml

```diff
@@ -0,0 +1,75 @@
+name: mint/build-leaf
+version: 0.0.0
+
+parameters:
+  dir:
+    description: 'Directory containing leaf contents'
+    required: true
+  rwx-access-token:
+    description: 'rwx access token with permissions to build a leaf'
+    required: true
+
+tasks:
+  - key: system
+    use: []
+    run: |
+      ran_update=0
+      if ! which zip; then
+        set +e
+        sudo apt-get install zip
+        apt_exit=$?
+        set -e
+        if [[ $apt_exit -eq 100 ]]; then
+          ran_update=1
+          sudo apt-get update
+          sudo apt-get install zip
+        fi
+      fi
+
+      if ! which jq; then
+        set +e
+        sudo apt-get install jq
+        apt_exit=$?
+        set -e
+        if [[ $apt_exit -eq 100 ]]; then
+        ran_update=1
+          sudo apt-get update
+          sudo apt-get install zip
+        fi
+      fi
+
+      if [[ $ran_update -eq 1 ]]; then
+        sudo apt-get clean
+      fi
+  - key: build
+    use: system
+    run: |
+      latest_timestamp=$(git ls-files -z ${{ params.dir }} | xargs -0 -n1 -I{} -- git log -1 --date=format:"%Y%m%d%H%M" --format="%ad" {} | sort | tail -n 1)
+      if [[ -z "$latest_timestamp" ]]; then
+        echo failed to determine timestamp
+        exit 1
+      fi
+      find ${{ params.dir }} -exec touch -t "$latest_timestamp" {} \;
+      leaf_name=$(grep '^name:' ${{ params.dir }}/mint-leaf.yml | awk '{print $2}' | cut -d '/' -f1)
+      leaf_file=/var/mint-workspace/$leaf_name.zip
+      echo -n $leaf_file > $MINT_VALUES/leaf-file
+      cd ${{ params.dir }} && zip -X -r $leaf_file .
+    outputs:
+      values:
+        - leaf-file
+  - key: upload
+    use: build
+    run: |
+      echo "leaf-file is ${{ tasks.build.values.leaf-file }}"
+      curl \
+        --request POST \
+        --fail-with-body \
+        --header "Authorization: Bearer $RWX_ACCESS_TOKEN" \
+        --header 'Accept: application/json' \
+        -F 'file=@${{ tasks.build.values.leaf-file }}' \
+        https://cloud.rwx.com/mint/api/leaves | tee leaves-result.json
```

Comment:
Should this be parameterized?

Reply:
Maybe. See the note in https://github.com/rwx-research/mint-leaves/pull/20 about publishing to staging.

I suspect for use cases of building leaves outside of production leaves, we're actually going to replicate the logic inside the leaf rather than using this leaf itself, because there's a bootstrapping problem of making this leaf available in the target environment. For example, if we want to test building and uploading a leaf to a local Cloud, we'd have to first get this leaf itself into the local Cloud. We could definitely do that via seed data setup, but I'm going to hold off for now.

Reply:
Ah, that makes sense. Thanks for the context. Missed that PR from earlier.